### ISSUE TITLE
Kestrel + IIS is falling over with a small load in production with RTM

### ISSUE THREAD

**Issue by jruckert** [aspnet/IISIntegration#245]
We are currently experiencing an issue in production (after upgrading from DNX to RTM) when our Web API is
- loaded with around 10 req/second
- Some calls are taking > 2 second with sometimes over 1mb in a payload. 
- Our average API timing and payload would be < 1 second and around 30kb of data.

It seems that sometimes when the larger API get queued whilst processing the smaller requests it overwhelms Kestrel.

We have a setup as follows:
- Dotnet Core RTM (1.0.0)
- Webhosting pack (1.0.0) installed
- Windows Server 2012 R2
- Large number of API's
- HTTPS
- Windows Authentication
- Entity Framework 7
- SQL Server 2014 (different server)

What we find is that we start eventually getting 502.3 errors and then Kestrel become unresponsive or takes a very long time to respond (e.g. > 30 seconds for a specific API when usually the API would respond in less than a second). 

The only way to resolve this issue is to recycle the app pool and kill the process thread (currently we are having to do this at least once a day in production).

Do we need to tune IIS to handle this scenario and how can we diagnose, provide log files to get to the bottom of this, as prior to RC2 (we upgraded from DNX through to RTM) the website has never crashed.

**Comment by jruckert** [comment]
We are also planning on running the load tool wrk for a few larger API's > 2 seconds and large JSON response payload to see if we can replicate this issue if this would help.

**Comment by jruckert** [comment]
From the IIS Performance Counters:

5:20pm - thread count started growing dramatically
5:40pm - access issues to website first reported
5:50pm - IIS Application Pool stopped and started, and thread count drops

Process counters for P… refer to the NetCore process.
Process counters for w… refer the IIS w3wp worker process.

![iisw3wp-threadcount](https://cloud.githubusercontent.com/assets/4064588/17362187/1d8e7490-59b3-11e6-8c84-bc4b442fe8a5.png)
![netcore-threadcount](https://cloud.githubusercontent.com/assets/4064588/17362188/2027ebb4-59b3-11e6-936d-15336b541699.png)

**Comment by wpostma** [comment]
I am having a similar problem.  Nowhere near that level of load.  3 tiny dotnet apps, on a server that also has an SQL Server 2016 instance.

I get about 6 seconds of 404 errors, followed by a permanent 502.5 condition, that lasts until I do a stop on w3svc and then start it again.  Restarting app pools is not sufficient.

http://stackoverflow.com/questions/38613615/dotnet-process-inside-iis-crashes-how-can-i-determine-cause-and-remediate/38745424#38745424


Do you get anything interesting if you enable stdout logging?

What do you see in the iis text log files (c:\inetpub\logs)? Is it similar to the log output in my answer in stackoverflow link above?

**Comment by KLuuKer** [comment]
Looks like i may also have the same issue,
we finished converting our framework 4.6 mvc 5 to core mvc 1
our traffic monthly is about:
- unique visitors 250.000
- uncached requests 1.300.000

Our setup:
- Dotnet Core RTM (1.0.0)
- Webhosting pack (1.0.0) installed
- Windows Server 2012 R2 (all patches installed)
- **HTTP** (tls is getting offloaded on a different machine)
- **Anonymous authentication** (public facing site)
- **NO** entity framework (this is the frontend, the backend exposes data via json\rest)
- SQL Server 2016 (different machine)


dotnet --version
```
Microsoft .NET Core Shared Framework Host

  Version  : 1.0.1
  Build    : cee57bf6c981237d80aa1631cfe83cb9ba329f12
```

```
Hosting environment: Production
Content root path: C:\inetpub\project.name-mvccore
Now listening on: http://localhost:21704
Application started. Press Ctrl+C to shut down.
```

and then all loads of variations of this that never get finished
```
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
      Request starting HTTP/1.1 GET http://example.com/variouspaths/that-should-work/
```

**Comment by KLuuKer** [comment]
@shanselman  @DamianEdwards 
sorry to @ you like this, but we are letting people hit our old mvc 5 app until this can be fixed
and so we cant deploy new code in the mean time (saas e-commerce app and our customers are getting angry)

**Comment by muratg** [comment]
cc @SajayAntony @sivagms

**Comment by muratg** [comment]
@KLuuKer Could you share the logs?

**Comment by pan-wang** [comment]
Could anyone provide a repro so that I can take a look at it?
404 error is generated by Krestral.
502 means IIS could not start the Krestral process. For this error, please take a look at windows event log. It usually gives some error code indicating the root cause of starting failure.
Please also try enabling IIS FREB log which might help to identify the root cause.

**Comment by KLuuKer** [comment]
@muratg can you give me your email here or dm via twitter to https://twitter.com/KLuuKer

**Comment by wpostma** [comment]
I have set up an indentical system to the one that is falling over on my W2012R2 vm, on an azure vm, but it's using Windows Server 2016 TP5 on the azure VM, and if it reproduces for me, I'll get access to any MS employee who needs it.  Right now I'm running exactly the same setup as the local VM that failed, except without using the IIS URL Rewrite extension.     If no reproduce, I'll add back in IIS URL rewrite.    I suspect that the problem is exacerbated or made more likely by me using three apps in a single pool.

**Comment by wpostma** [comment]
@pan-wang  How do I enable IIS FREB for this context?   I don't know what to enter for "Specify Content to Trace" prompt in iismgr. All contents? Then 502 error?
![freb](https://cloud.githubusercontent.com/assets/1833470/17416380/da76c082-5a5b-11e6-988a-4eb40c279197.png)

**Comment by KLuuKer** [comment]
Just making this clear: i am running kestrel\mvc core 1 using the full framework 4.6 because of legacy bits

**Comment by jruckert** [comment]
Also the spin up time for our process due to the large dependency injection requirement is approx 5 seconds before IIS can access it. Would this affect things?

**Comment by pan-wang** [comment]
IIS will log the start and failure of Kestrel process to windows event log. In case of 502 error, this log will be most important source to help identifying the root cause of the failure.  If possible, please do take a look at it and share it.
 
@jruckert IIS has a timeout setting on Kestrel process starting, default is 2 minutes. 
@wpostma Yes, you can choose all contents and your current FREB setting looks good. For your reference, current hosting mode is one Kestrel application per IIS apppool (site). There will be some issues especially about recycling/file change if you host multiple apps together. I am working on changing it to allow user to host multiple applications inside one apppool.

**Comment by GuardRex** [comment]
Ping @tuespetre ... and I ... have an interest in your (@pan-wang) comment on multiple apps per pool, as @tuespetre is doing it in staging, and the green dino here is :chicken: too chicken :chicken: to try it at all. :smile:

> I am working on changing it to allow user to host multiple applications inside one apppool.

@pan-wang, there is no GH issue we can track on for that, correct?

**Comment by KLuuKer** [comment]
From what i have seen on my server, sometimes it starts up correctly getting hit with real user traffic and after a while it starts getting slower, later it just slows to a crawl and then just stops and i get timeouts

If at that moment I recycle the apppool or manually stop iis (iisreset /stop) and start iis (iisreset /start)
i still keep getting timeouts

when it starts giving timeouts (before and after restart) i also notice kestrel using almost no cpu (1% and less) and the entire cpu also running at idle doing almost nothing

**Comment by KLuuKer** [comment]
@jruckert @wpostma do you also some code like mine above?
or just normal async await code with allot of HttpClient calls to a different machine?

**Comment by jruckert** [comment]
I'm having a similar experience where it seems to gets slower but it also sometimes recovers. As we are only hosting Web api controllers I'm not 100% sure if all client calls are using the approach you have outlined.

**Comment by KLuuKer** [comment]
TL;DR its **not** the end users\browsers that are using that code i am

that is the code that i have in my frontend server (kestrel\iis) that is getting hit with browser requests
and since i have a n-tier system the frontend (my code) is using that piece of code to hit a different machine my backend server (web api 2\mvc5\iis\EF6) and that grabs it from the database and returns it

**Comment by KLuuKer** [comment]
Is there some way of limiting the amount of concurrent request in iis?
I would like to try to see if limiting the amount of simultaneous request can temporarily fix the problem.

**Comment by KLuuKer** [comment]
so i removed that old  `return Task.Run(async () => {  // work here  }.Result; ` code and......
it did NOT help as soon as the users started hitting the site again hard it began to crawl to a halt again
so no luck there

**Comment by GuardRex** [comment]
@KLuuKer You might get some good info if you PerfView the app before and after it goes into :snail: mode.

**Comment by davidfowl** [comment]
@KLuuKer IIS limits concurrent requests to 5000 by default.

Has anyone collected a profile or process dump? Where is the CPU time spent? Also is we shine using IIS + Kestrel?

**Comment by muratg** [comment]
Folks, could someone also try to use a single app per app pool to see if this is the culprit?  @pan-wang is working on better supporting this in a future release of AspNetCoreModule, but perhaps this could be the work-around for now.

**Comment by KLuuKer** [comment]
@davidfowl yes i am using IIS (win 2012r2 fully patched) + kestrel
anyway to lower that (dont mind machine wide) but keeping the queue big?

@muratg i am using **single** app per app pool

also the strange thing is i dont see big cpu usage (or 25% because its a quadcore box) more like below 10% total cpu usage

**Comment by Tratcher** [comment]
Can anyone share an app that reproduces the issue for us to investigate directly?

**Comment by KLuuKer** [comment]
i tried making a repro app but failed (will try again)

@wpostma how is your repro going?

**Comment by pan-wang** [comment]
This post lists a way to viewing long running IIS request inside InetMgr https://technet.microsoft.com/en-us/library/cc732518(v=ws.10).aspx. You can also use perfmon to monitor IIS perf counters such as Active Requests under W3SVC_W3WP. The max number of active requests is 5000. Once this max is reached, IIS will not server new quest. Could you please open perfmon to take a look at this IIS perf counter?

**Comment by jruckert** [comment]
I'm still working on a repro and using wrk as the load simulator.

**Comment by KLuuKer** [comment]
@jruckert are you able to repro it using wrk on the full app? and are you now just cutting down the code to where the problem is?

**Comment by jruckert** [comment]
I think it is a combo of a couple of simultaneous  heavy calls with a couple of short calls.  I'm in the process of rerunning the IIS logs as this is our best chance of reproducing. We are hosting approx 1200 api's and the consumer base is well over 20 million so the sooner we get to the root cause the better.

**Comment by jruckert** [comment]
BTW I'm unfortunately unable to simulate this in production every day due to the existing load failing over every day. I have no roll back plan as we have committed to dotnet cli 12 months ago.. So I'm reliant on the IIS logs to replay load. As we are publishing government records back to 1800 this is a major issue.

**Comment by davidfowl** [comment]
Does anyone have a process dump? That might help as well.

**Comment by KLuuKer** [comment]
I took another try at running our production traffic trough kestrel\mvc core
but it crashed again here are some things i noticed

cpu usage on sql, backend, frontend (this runs mvc core) machines is below 10%
tcp connections on the frontend are below 2000 total (including those waiting for timeouts) so no ephemeral port are exhausted

it seems to be a combination of request, just allot of GET requests to / (the frontpage) is not enough to replicate the problem

**Comment by DamianEdwards** [comment]
@KLuuKer can you paste your `Startup.cs`, `project.json', & `web.config` files here? Pls ensure you're not using the Console logger in production, as that takes a global lock on every log message. Configure something like Serilog to log to a text file for in-app logging and just have startup errors go through console out to the configure txt log via `web.config`.

**Comment by KLuuKer** [comment]
@DamianEdwards you are not going to like this, i don't either and i want to clean it up eventually (you know when i will get "around" to do it).

i should be on gitter when you read this message

web.config - https://gist.github.com/KLuuKer/2ddc14ee76c259efdd020a9e76cb204e
program.cs - https://gist.github.com/KLuuKer/d5da65927244174a70083eb8a4334b58
project.json - https://gist.github.com/KLuuKer/d1f5106f941ef80ec472d2f8f012e08b

and please make sure your seat is in the upright position the: "omg what the hell are your doing please do this in a different way for the love of programming"
startup.cs - https://gist.github.com/KLuuKer/95d0b76f56ea29bd56caeeb585726929
(feedback welcome)

**Comment by jruckert** [comment]
@davidfowl We have a process dump of the issue how can we get this to you??

**Comment by KLuuKer** [comment]
@DamianEdwards i switched to Serilog and removed console log, its still has the same 0% cpu and a ton of request pouring in getting timed out

**Comment by muratg** [comment]
@jruckert Could you upload it to onedrive or similar and share a link?

**Comment by jruckert** [comment]
@muratg I am assuming this process dump has the potential to contain confidential customer information. I can put up on a secure download site.  Can someone DM me at twitter.com/jruckert

**Comment by muratg** [comment]
@jruckert Couldn't send a DM because you're not following me on Twitter. :)  However, you can email me at mgirgin  @  microsoft  .  com.

**Comment by pan-wang** [comment]
@KLuuKer did you check IIS active request counter when the hanging happened?

**Comment by KLuuKer** [comment]
@DamianEdwards tested again this time also removed\disabled the timers and the websocketclient i have in there, same problem

@pan-wang request monitoring looked like this 
![request-monitoring-privacy-stripped](https://cloud.githubusercontent.com/assets/12729178/17583262/4b7a935e-5fb0-11e6-9002-44e2aa69996c.png)

**Comment by jruckert** [comment]
I have been able to reproduce the issue consistently within our test environment and using wrk as the load tester. As it's a large code base what would you recommend the next step is to provide a public repo?

**Comment by jruckert** [comment]
It is definitely a load issue and related to a long running query.  We see that the dotnet process hovers around 100% CPU after about 5 seconds although SQL is only running at 5% after an initial burst. It seems that it's still trying to process something but it stops calling SQL after a while.

**Comment by KLuuKer** [comment]
@jruckert so with your problem the process is stuck at 100%? because mine is that its stuck at 0% and it looks like its waiting for some lock somewhere

yeah public repo would be very helpful because Microsoft can then do all the profiling and dumping they want, so start stripping out some of the code that is not touched, repo again, make a backup of the stripped version, and keep stripping\repoing\backuping until you cant repo

**Comment by KLuuKer** [comment]
today we tried running on azure windows server 2016 TP 5 that also didn't help
we tried with a quad core and a single core machine

on the quad core it looked good but after a hour or so
and the single core machine was good for 2 hours

the both ended up taking really long for each request and they started to timeout
again with next to nothing cpu usage (total usage below 5%)

**Comment by jruckert** [comment]
Also what should i expect when running the following:

wrk -t5 -c50 -d60s https://webapithatisquick

wrk -t5 -c50 -d60s https://webapithatisslow

When i run the top command, the site works fine

When i run the bottom command, the site hangs and does not recover unless i recycle the app pool.

Would this be expected behaviour?

**Comment by jruckert** [comment]
@muratg Is it best that I log a premier support ticket on this as the customer is not the happiest in the world at the moment.

**Comment by davidfowl** [comment]
@jruckert You should do that as well. Github isn't the same as customer support. We're looking into it but you should use as many channels as you have at your disposal.

**Comment by davidfowl** [comment]
> It is definitely a load issue and related to a long running query. We see that the dotnet process hovers around 100% CPU after about 5 seconds although SQL is only running at 5% after an initial burst. It seems that it's still trying to process something but it stops calling SQL after a while.

Are you using EF?

**Comment by pan-wang** [comment]
@KLuuKer 
Could you please take a look at the Active Requests counter like this one?
![image](https://cloud.githubusercontent.com/assets/18172635/17604243/d1097e5e-5fc9-11e6-838e-66184e204a28.png)
If the active request counter value is 5000 and w3wp does hang, it could be a bug in IIS code for your scenario. I need a either a dump or a repro to investigate the root cause.

**Comment by jruckert** [comment]
@davidfowl yes we are using EF heavily.

**Comment by davidfowl** [comment]
> It is definitely a load issue and related to a long running query. We see that the dotnet process hovers around 100% CPU after about 5 seconds although SQL is only running at 5% after an initial burst. It seems that it's still trying to process something but it stops calling SQL after a while.

@jruckert do you have any idea what query is? If you remove those queries, do things start working?

/cc @rowanmiller

**Comment by jruckert** [comment]
Yes we are only hosting Web api's so we can determine the query but I know some are very complex.  The long query (join and grouping)  from the Web api  takes about 4 seconds.  If we use a more simpler e.g. quicker query less than a second then we can load it up more so.  I'm not sure if the EF is processing a lot of things in memory though this is something that I will have to chase down.

**Comment by jruckert** [comment]
Below are some stats for stress testing the API /api/LongRunningApi

Note that a single call (non-optimized) took about 10 seconds and used 50% of available CPU on a server with dual core 2.1GHz processors.

The Web.Config is configured with a RequestTimeout of 120 seconds.   However, the NET Core process keeps running (Duration NET Core process was stressed) even after the IIS timeout, so there is no way for it to return any results.

From this stress testing and other observation, it appears the NET Core processes keep running until they finish, even though there is no way for them to return any results to IIS, AND may overwhelm the CPU the whole time.
•	With the stress test for 40 concurrent requests, it kept running for about an extra 200 secs (260-60) after the requests were no longer being sent.   Assuming an even distribution of how far each of the 40 requests had processed, total remaining execution time will be approximately half, so 10secs/2 * 40 requests = 200 seconds, which matches.
•	Last night, I estimate the NET Core process was using 100% CPU for over 30 minutes before it recovered, and no extra requests were sent during this 30 minutes.  This was for a stress test for several hundred requests.

Sorry about the table formatting..

Concurrent API Requests	CPU
(2 cores)	Duration of stress test	Calls completed during test	Duration NET Core process was stressed	Time for /api/SimpleApi (this is a very basic SQL statement) call to load during the test
BEFORE CODE OPTIMIZATION	 	 	 
1	50%	60 seconds	6	70 seconds	fast sub-second
2	98%	60 seconds	12	70 seconds	fast sub-second
4	100%	60 seconds	9	70 seconds	fast sub-second
5	100%	60 seconds	10	80 seconds	1 - 2 seconds
10	100%	60 seconds	7	90 seconds	over 10 seconds
20	100%	60 seconds	8	150 seconds	over 20 seconds
30	100%	60 seconds	7	190 seconds	44 seconds
40	100%	60 seconds	3	260 seconds	timeout after 2 minutes
AFTER CODE OPTIMIZATION	 	 	 
40	80%	60 seconds	251	90 seconds	4 seconds
200	80%	60 seconds	347	90 seconds	33 seconds

Our code optimization was:

changing a single line of code from DbSet.Where to DbSet.AsNoTracking().Where

**Comment by KLuuKer** [comment]
So it looks like my slow request for images (that need to be resized) are causing all other requests that are fast to wait, and to fill up the request queue

**Comment by KLuuKer** [comment]
yes it is definitely a bunch of slow request making all the other request wait in line to get processed
i have put a simple cache around my image resizing logic and when its all warmed up it runs super fast 
but get 60~90+ active request's and o boy it slows to a crawl

**Comment by Tratcher** [comment]
Now that we have some idea of the scenario, have you tried it again with just Kestrel? We'll see if it's an issue with Kestrel or IIS/ANCM.

**Comment by KLuuKer** [comment]
@Tratcher will try tomorrow UTC+01:00 so no time for fun stuff today

**Comment by veccsolutions** [comment]
I thought I would mention (and add more information) that I too am having the same issues but with a slightly different setup. The 502's will stop after the long running requests stop and I don't need to restart the app pool. While it's paused I get random 404's. I'm able to reproduce it on demand, I just have to send in about 30-40 long running requests at the same. Not sure if it matters, the solution was originally RC1, RC2 then finally upgraded to RTM. I don't use EntityFramework, and instead use Dapper. I use netstandard16 as my framework, IIS on server 2012R2. I also make a soap request to a backend server which is where most of the time is spent. Everything is done using async/await for network resource access (database & soap)

The timings I have for my service (from start of first request to last byte of last request):
1 request 500ms
2 requests 500ms
4 requests 1 second
8 requests 1.4 seconds
16 requests 1.5 seconds
32 requests 15 seconds (with multiple 404's being returned) - when there are no 404's/502's its about 2.5 seconds

**Comment by muratg** [comment]
We're looking at @jruckert's process dump, though in the meantime, could someone try and see if they can reproduce the issue without AspNetCoreModule (i.e. without IIS)?

**Comment by veccsolutions** [comment]
While running the solution without IIS I was able to reproduce the same timings that I posted but not the 404's and 502's. I'm not sure if this helps, but it's more information. While it's taking forever, pressing Ctrl+C doesn't kill the process.

**Comment by davidfowl** [comment]
Ctrl+C lets requests unwind gracefully for some period of time before tearing things down. So if you have long running requests it would hang for a little.

**Comment by veccsolutions** [comment]
Good to know. Would that be why sometimes when I stop the app pool the dotnet process doesn't terminate right away?

**Comment by muratg** [comment]
cc @yuwaMSFT

**Comment by wpostma** [comment]
I have successfully reproduced a similar issue, and have created a VM on azure and can give credentials to log in remotely to anybody from MS who wants to have a look.

1.  created one IIS app pool and deployed MULTIPLE apps to the same pool because it seemed the only way to make url rewriting work.
2.  The apps work, although they don't auto start, they start manually when you access them via their non-rewritten points.
3.  Using a python script I continuously do http gets to various endpoints on the three .net core 1.0 services.
4. After some time, between 3  hours and 10 days, of continuous access about 100 times a second (so not very heavy load), it tends to lock up, and api endpoints give 502.5 error instead of the proper responses (such as 200 OK, or 401 unauthorized) that my test script expects.

5. Recovery requires stopping and restarting IIS, and restarting an app pool is not sufficient.

6. On the vm itself, I have made various notes and references which are all visible if you browse http://localhost/

**Comment by natemcmaster** [comment]
We've done some investigation of the repros and memory dumps. So far, we haven't found one single bug causing IIS + Kestrel to fall down.  Rather, it appears the downfall is 'death by a thousand papercuts'. 

I've gathered a list of some of the issues we've found, some already alluded to in responses above. The issues range from user code to low-level IIS issues. This is by no means an exhaustive list of performance tips, but is a good place to start when investigating perf issues in your individual apps.

### ASP.NET Core Module request timeout
ANCM will timeout with a 502.3 error if the backend process (dotnet.exe) hasn't begun writing a response after 120 seconds (the default request timeout). Users may notice the dotnet.exe process continues at high CPU even after the IIS worker process has killed the request. ASP.NET Core will not kill a running request even if the client has hung up.

In Kestrel 1.0.0, `HttpContext.RequestAborted` is not triggered until the response attempts to write to an aborted request. With https://github.com/aspnet/KestrelHttpServer/pull/953, this token will be triggered when ANCM hits the request timeout.

**Suggestions**

 - Use `HttpContext.RequestAborted` to terminate requests early. 
 - Increase the ANCM request timeout in web.config https://docs.asp.net/en/latest/hosting/aspnet-core-module.html
 - Make your long running requests faster (... in case that wasn't already an obvious solution).

### Console logger
Under high load, the locks used in console logger may reduce performance of a web app. 

**Suggestion**
 - Use something like Serilog instead of 'stdout > file' for production app logging.

### EF Core change tracking
By default EF Core will use the change tracker to snapshot and track all entities materialized from the database. This is a common source of memory and CPU waste in web endpoints that only query data, but don't modify it.

**Suggestion**
 - If the app does not need change tracking, turn it off by using the `.AsNoTracking()` call in LINQ queries.

### App pool recycle starts app in wrong directory
In setups using an application nested in a folder other than the IIS website root, app pool recycling may attempt to start ANCM in the wrong directory, using the website root directory instead of the application directory. This occurs when the server is under high stress. If web.config uses relative paths, the backend process for the web app will not start correctly, and may emit a logger error such as

> The application to execute does not exist: 'C:\inetpub\websiteroot\testapp1.dll'

ANCM will return 502.5 errors as long as it is unable to restart the backend process. IIS will continue to attempt to restart the app, but may not succeed, depending on factors we are still investigating. 

Note: this may also be the cause some 404 errors, depending on the IIS website structure.

**Suggestion**

 - Tune app pool recycling settings the recycle during low load.

**Comment by wpostma** [comment]
Does IIS try more than once to restart the app or having picked the wrong working directory does it sit there and not retry?

**Comment by pan-wang** [comment]
ANCM will try to start Krestrel for every request if the process has not been started yet. I looked at the log and noticed that a wrong application physical path was used for starting Krestrel for some cases under stress. This may relate to your scenario that you use url-rewrite to redirect the requests which will recreate a child request. Under stress, IIS may resolve wrong root context to the parent request and thus cause wrong physical path would be used.  I have updated ANCM code and have a private ready. Could you please give a try?    

Note that ANCM starts Krestrel in suspend mode first in order to assign job object, resumes thread and then calls GetExitCodeProcess to see whether the process is still running. If Kretrel returns value other than STILL_ACTIVE, ANCM will treat it as failure.  I did see some cases that Krestrel did not return STILL_ACTIVE.

**Comment by veccsolutions** [comment]
Well, I found a workaround, at least the hanging stopped anyways. In program.cs, where we call .UseKestrel, I set it to use 1000 threads for the kestrel server. 1000 is way more than I will need at a given time and haven't tried lower values. I have yet to get it to hang after that change. I'm now pushing 100+ requests at a time with no issues or slow downs. Digging through the code it looks like it will only use the same number of threads as processors/2. Is that correct?

Here's my updated webhostbuilder.

        var host = new WebHostBuilder()
                .UseKestrel((options)=>
                {
                    options.ThreadCount = 1000;
                })
                .UseContentRoot(Directory.GetCurrentDirectory())
                .UseIISIntegration()
                .UseStartup<Startup>()
                .Build();
`

**Comment by KLuuKer** [comment]
Even with 1.0.1 and the ThreadCount set to 500 it still happens in production
i am more then happy to try beta builds

**Comment by pan-wang** [comment]
@KLuuKer  I have an ANCM private addressing the child request and some other issues. This private may help if you used urlrewrite module. If you want to give a try, please let me know.

**Comment by KLuuKer** [comment]
@pan-wang  I am not using the urlrewrite module, but i would love to give it a try non the less
you can DM me on https://twitter.com/KLuuKer for the link
any specific instructions about installing the private and installing a new public version when it comes out?

**Comment by halter73** [comment]
@KLuuKer did you ever get around to testing the app without it being behind IIS? Are you able to get a memory dump from the application process when it is behaving poorly?

**Comment by KLuuKer** [comment]
@halter73 i grabbed a azure vm and let our production load hit kestrel directly but to many request were slow or just took to long
so we had to abort it

**Comment by shanepask** [comment]
We had this issue and it wasnt IIS related, we managed to recreate by holdind Ctrl+F5 on chrome.

Our setup was to call an API service using an HttpClient, this seemed to create persisted connection to the API that caused it to grind to a halt and the thread count to rise.  

Investigation revealed that the underlying connections seem to remain open until a response was generated even if the connections on the caller where closed and GC called on the caller.

Disposing the HttpHander and HttpClient had no effect on the API receiving a ECONNECT error instead the request pipe just continued to fill with unfinished concurrent request.

Cant confirm this but looked like the CPU was having trouble with moving between threads, profiling relieved a lot of WaitHandle time.

With HTTPCLIENT, Possible solution... 

please try and let me know if this fixes anyone else issue with similar setup.

Is anyone else using an HttpClient, if so adding....

```
            resp.RequestMessage.Dispose();
            resp.Dispose();
```
Onto the HttpResponseMessage object has freed up the API and requests seem to be cancelling when the connection is closed.

Don't wanna get my or anyone else's hopes up too high as i can only say this issue has be a journey that i can believe *may* be over.

**Comment by KLuuKer** [comment]
@Tratcher  @DamianEdwards @pan-wang @halter73 guys can we please get some private unsupported beta builds, we are recycling the app pool every 30 minutes its getting crazy

**Comment by wpostma** [comment]
I had a strange "IIS falls over" occur today where the entire app pool broke at startup, and the app pool didn't even start.

**Comment by halter73** [comment]
https://dotnet.myget.org/F/aspnetcore-ci-dev/api/v3/index.json Is the nightly feed. A memory dump from when an app is in a bad state would be helpful for finding the root cause.

**Comment by pan-wang** [comment]
@shirhatti could you please help to pass the ANCM private bits?

**Comment by KLuuKer** [comment]
@halter73 i have a dump with about 100+ active requests (iis perf counter)
send me a DM on https://twitter.com/KLuuKer 

i will give a onedrive link and a password for the zip that has the dump

**Comment by halter73** [comment]
@KLuuKer I sent you a DM. Did you see it?

**Comment by KLuuKer** [comment]
@halter73 yes, gave you a reply
sorry i didn't see it sooner twitter isn't great at showing you have unread messages

**Comment by KLuuKer** [comment]
@halter73 please keep the dump need to know as it can contain private data that should not be shared

**Comment by halter73** [comment]
@KLuuKer Do you have a VS ultimate license? If so, I highly recommend taking a look at your minidump using [VS's Debug Managed Memory feature](https://msdn.microsoft.com/en-us/library/dn342825.aspx?f=255&mspperror=-2147217396#Anchor_4). If not, you can use WinDbg and Sos with the !dumpheap and !clrstack commands to get the same info, but far less conveniently.

If you do this, I would first take a look at the "Parallel Stacks window". This will show you have 595 managed threads currently executing (out of 607 total native threads). If you search the heap, for instances of an object that is created once per request (like say DefaultHttpContext), you'll see 334 instance meaning that there were more than 334 active requests when the minidump was taken. Looking at some more internal types I think the number is closer to 231 active requests with some of the other HttpContexts still being rooted outside the context of a request, but 334 is certainly an upper bound.

By far most of the threads, 437 to be exact, are blocked waiting on Tasks inside a method called CompanyState.TimerTick. Another 50 threads are blocked waiting on Tasks in a Action/lambda defined inside LocalCacher.CacheLookupLongWithRenew. And another 40 threads are blocked waiting on Tasks in IpState.GetFromDb.

Most of the outstanding tasks seem to be indirectly caused by the threadpool starvation. By looking at the ThreadPoolWorkerQueues and WorkStealingQueues, I was able to calculate that there were over 1100 pending threadpool work items waiting for threads to become available. 767 of these work items are Tasks waiting to complete (which can be seen by inspecting the ThreadPoolTaskScheduler object). All 767 of these Tasks have essentially completed, but are still in the [WaitingToRun](https://msdn.microsoft.com/en-us/library/system.threading.tasks.taskstatus(v=vs.110).aspx#Members) state since the threadpool is starved. About 437 of those Tasks are `Task<Task<Models.Company>>` and another 50 are `Task<Task<Models.BannedI>>` which are the tasks that the TimerTick and GetFromDb methods are blocked waiting on.

You wound up in a situation very similar to a deadlock where one thing is using a limited resource (being a threadpool thread) while waiting on another thing that needs that very same resource to complete. It’s not a true deadlock since more threads will be slowly added to the threadpool which would eventually drain as long as no more external work is added, but you can see why it problematic.

**Comment by KLuuKer** [comment]
@halter73 I would have never guessed that my code would get "deadlocked" by threadpool starvation.
I have made some emergency hotfixes to the code and it's running super now, no more "hangs" with 0% cpu.

The Parallel Stacks window one of the many many options in visual studio i totally forgot about, yes looking at it made it very clear where the problem was in my code.

Thank you very much for the great write up, I think this kind of issue will pop up more as other (legacy) apps get converted to use all the async goodness. 

I'm very sorry to all the people I poked and bugged about my own buggy code, thank you very much everyone for helping me find this. I cannot wait to move my backend-tier over to .NET core, this time making sure I don't create so many tasks :smile:.

**Comment by csharmath** [comment]
@KLuuKer @halter73 @DamianEdwards @davidfowl 

First of all, really interesting thread - love to see how it unfolded and resolved. 

I've seen struggles like this many times in various situations and it feels like that the analysis of IIS/ASP.NET hang or high CPU is often seen as black magic requiring "complex / hard to remember" tools such as windbg so many won't do it / won't invest the time they probably don't have to master it.
( not that I agree as I think that's a very good investment for any dev)
So my point is that I think this should be made way much more simpler for an app dev to identify that there is a thread starvation issue. in production, on a server without VS in less than 2 minutes.

I don't often see for example PerfView recommended, but it has some really awesome features to analyze ETW traces and you can drill into ASP.NET, etc.
Sometimes you don't even need windbg to tell what's wrong. ( I wonder if it would have helped in this particular case )

If it's the right tool, then PerfView should be promoted more or the community could use even more simpler tools to very quickly and obviously tell if it's a thread contention / starvation issue.

The point is that it seems that it's way too easy and frequent to make the typical thread blocking mistakes but it's not as obvious and easy to quickly and simply identify them if you are new to these or haven't done this production scaling firefights.

Performance and scaling is not something which should be an afterthought ,but it should be part of the design and tools should aid or guide devs to be able to get it right. but in case it wasn't the case then still we should make it dead simple to tell that this app can't scale by design.

my 2 cents on the topic, let me know what do you see from msft side or how others see this

Thanks!

**Comment by davidfowl** [comment]
An analyzer might help identify some bad coding patterns earlier as well (like using Task.Result 😄).

**Comment by robertmclaws** [comment]
@shanepask FYI HttpClient is a static instance internally, so if you are not reusing the instance, or calling Dispose() on it (as all Microsoft's docs tell you to do, so don't feel bad), you're likely running into this issue: http://aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/

How we solved it was to use DI to inject a shared HttpClient instance into all of my controllers. Then, before I use it, I check to make sure it's still around, in case it got accidentally disposed of.

HTH!

**Comment by gencebay** [comment]
I have a example repo for similar issue we encountered
[Net Core Small Load](https://github.com/gencebay/NetCoreSmallLoadRepro)
Thanks!

**Comment by mupedersen** [comment]
We had a similar issue in an aspnet core (.net full framework) solution. Calling the application start page from **two** machines would degrade performance to a very low level - fx loading css, js and image bundles would take +60 seconds. Once it was in the degraded performance state, it would stay there until running: 

net stop was /y
net start w3svc

The start page only does a single db lookup (using Dapper) and loads 4 static files - all in all 305KB.  

Upgraded from 1.0 to 1.1 and the problem disappeared.

**Comment by krswope** [comment]
@KLuuKer 

> I have made some emergency hotfixes to the code and it's running super now, no more "hangs" with 0% cpu.

Can you give an example of something that you changed that cleared this up for you like "change this to this?"  I am having a similar issue where IIS is just getting locked up at around 500 concurrent users on my api and then it just dies.  Anything you can share would be amazing.  Thanks!

**Comment by KLuuKer** [comment]
@krswope i recommend you try the managed debugger feature in VS used by halter73  https://github.com/aspnet/IISIntegration/issues/245#issuecomment-253668589
it is really easy to track the thread pool starvation problem with

i had timer's ticking that started async HttpClient requests, so i just increased the interval they where ticking at.
after that worked i started to build a better way that didn't require me running a timer

**Comment by krswope** [comment]
@KLuuKer thanks for the response.  I'm using VS 2015 Enterprise so that Managed Memory Feature is different.  I'm having trouble trying to nail down where the issue is happening.  Any tips on that would be great.  I have some stopwatches that I kick off before like this:
`apiWatch.Start();
 HttpResponseMessage response = await client.GetAsync(getwhatever);
 apiWatch.Stop();`

Perhaps I should take these out.

**Comment by KLuuKer** [comment]
@krswope sorry i meant the "Parallel Stacks window"
it shows all of the stack's in parallel grouping by the stacks that are the same with a count

**Comment by muratg** [comment]
We are closing this issue because no further action is planned for this issue. If you still have any issues or questions, please log a new issue with any additional details that you have.

**Comment by 1amirjalai** [comment]
can someone please tell me how to create a dump file?
should i use adplus so on what condition should the dump be generated? 
what other tools can i use for generating dump and examining them?

**Comment by muratg** [comment]
@1amirjalai, on Windows?  Please take a look at https://technet.microsoft.com/en-us/sysinternals/dd996900.aspx

**Comment by foragerr** [comment]
@veccsolutions [This page](https://docs.microsoft.com/en-us/aspnet/core/api/microsoft.aspnetcore.server.kestrel.kestrelserveroptions#Microsoft_AspNetCore_Server_Kestrel_KestrelServerOptions_ThreadCount) also says threadcount is clamped between 1 and 16, which probably means setting it to 1000 only gives you 16 threads?

**Comment by BKlippel** [comment]
Apparently windows sucks, try Linux, where .net core can run up to 20x faster...
